checkers/deferAtTheEnd_checker.go

```diff
@@ -0,0 +1,81 @@
+package checkers
+
+import (
+	"go/ast"
+
+	"github.com/go-lintpack/lintpack"
+	"github.com/go-lintpack/lintpack/astwalk"
+	"github.com/go-toolsmith/astfmt"
+)
+
+func init() {
+	var info lintpack.CheckerInfo
+	info.Name = "deferAtTheEnd"
+	info.Tags = []string{"diagnostic", "experimental"}
+	info.Summary = "Detects calls to defer at the end of function"
+	info.Before = `
+func() {
+	defer os.Remove(filename)
+}`
+	info.After = `
+func() {
+	os.Remove(filename)
+}`
+
+	collection.AddChecker(&info, func(ctx *lintpack.CheckerContext) lintpack.FileWalker {
+		return astwalk.WalkerForFuncDecl(&deferAtTheEnd{ctx: ctx})
+	})
+}
+
+type deferAtTheEnd struct {
+	astwalk.WalkHandler
+	ctx *lintpack.CheckerContext
+}
+
+func (c *deferAtTheEnd) VisitFuncDecl(funcDecl *ast.FuncDecl) {
+	ast.Inspect(funcDecl.Body, func(n ast.Node) bool {
+		funDecl, ok := n.(*ast.BlockStmt)
+		if ok {
+			c.checkDeferBeforeReturn(funDecl)
+		}
+		return true
+	})
+}
+
+func (c *deferAtTheEnd) checkDeferBeforeReturn(funcDecl *ast.BlockStmt) {
+	retIndex := len(funcDecl.List)
+	for i, stmt := range funcDecl.List {
+		if _, ok := stmt.(*ast.ReturnStmt); ok {
+			retIndex = i
+			break
+		}
+	}
+	if retIndex == 0 {
+		return
+	}
+
+	if deferStmt, ok := funcDecl.List[retIndex-1].(*ast.DeferStmt); ok {
+		c.warn(deferStmt)
+	}
+}
+
+func (c *deferAtTheEnd) warn(deferStmt *ast.DeferStmt) {
+	s := astfmt.Sprint(deferStmt)
+	fnlit, ok := deferStmt.Call.Fun.(*ast.FuncLit)
+	if ok {
+		// To avoid long and multi-line warning messages,
+		// collapse the function literals.
+		s = "defer " + astfmt.Sprint(fnlit.Type)
+		if fnlit.Body.List == nil {
+			s += "{}"
+		} else {
+			s += "{…}"
```

Comment:
TBH I would like to omit Unicode characters as stick to simple ASCII.

Reply:
OK, I'll fix it

Reply:
Fixed